keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -231,7 +231,177 @@ TBD
 
 ### Implementation Details/Notes/Constraints
 
-TBD
+##### Definitions
+
+- *BucketBinding*: a set of rules defining how we bind an inbound request w/ a bucket.
+It works similar to \[Cluster\]RoleBinding in the RBAC world.
+- *Quota*: a request must acquire one quota before its execution. Basically,
+we have two kinds of quota: __reserved__ and __shared__. With reserved quota, the
+request can be executed immediately. A bucket's reserved quota is exclusively
+consumed by the request matches the bucket, while its shared quota can be consumed
+by the request matching higher or the same priority.
+- *Bucket*: a quota pool. The quota will be taken from the bucket when it's
+acquired by any request and will be returned to the bucket when the request
+finishes. A bucket has following attributes:
+  - *Priority*: a fixed/preset set of priorities ranged 1...N. The lower
+  number means higher priority. Each priority has one or more buckets providing
+  quotas and buckets from higher priority can "borrow" shared quotas from
+  the lowers.
+  - *Reserved Quota*: an initial guaranteed concurrency can be provided by
+  the bucket.
+  - *Shared Quota*: an initial shared concurrency can be provided by the bucket.
+  Shared quota cannot be consumed by requests queued at logically higher
+  (or numerically lower) priorities.
+  than the bucket's.
+  - *Weight*: only works for shared quota. The higher weight a bucket has,
+  the more chance the requests hitting the buckets can be executed from the
+  queue. Note that a zero weight means that the requests matches the bucket
+  cannot be executed by shared quotas, or rather, it can only be executed
+  by reserved quotas.
+
+- *Extra Shared Bucket / ESB*: a global logical bucket only providing shared
+quotas at a logical "lowest" priority (which is unreachable via external
+API requests). Requests doesn't match any priority will be queued in this
+logical "lowest" priority until the extra bucket distributes its quota to it.
+The extra shared quota can be consumed by requests from all priorities b/c
+it lies at the "lowest". In a high level, ESB defines a default common concurrency
+in the system and it's supposed to be auto-tunable by a moving average observed
+by the system. Note that requests queued in this extra bandwidth are popped
+strictly in a FIFO order.
+
+##### Bandwidth Goal
+
+We can define __Shared Bandwidth__ as a goal for priority band __p__, and
+we have N fixed priority:
+
+- *Shared Quota(p)*/ SQ: sum of shared quotas provided by
+buckets at Priority p.
+- *Total Shared Quota*/ TSQ: sum\[ SQ(i) over 1..N \] + ESB.sharedQuota
```

Comment:
The pedant in me likes to see variables bound.  An expression more like "ESB.sharedQuota + sum[over i in 1..N] SQ(i)" would go down smoother.